### PR TITLE
cam6_2_006: Update CLUBB/SILHS

### LINKED ISSUE #43: Updates to CLUBB/SILHS
. Modify SILHS to address conservation issues.
. Move CLUBB PDF parameters out of physics buffer to address a performance
  problem identified by John Dennis.
. Answers change for SILHS configurations only.

### PR DESCRIPTION
**Pull request by brian-eaton**

### CODE REVIEW

**Comment by brian-eaton** [comment]
This address issue #43

File: src/physics/cam/micro_mg_cam.F90 [hunk 1]
```diff
@@ -3282,6 +3368,20 @@ subroutine micro_mg_cam_tend_pack(state, ptend, dtime, pbuf, mgncol, mgcols, nle
    if (qsnow_idx > 0)  qsout_grid_ptr = qsout_grid
    if (nrain_idx > 0)  nrout_grid_ptr = nrout_grid
    if (nsnow_idx > 0)  nsout_grid_ptr = nsout_grid
+   if (qcsedten_idx > 0)  qcsedtenout_grid_ptr = qcsedtenout_grid
+   if (qrsedten_idx > 0 .and. micro_mg_version > 1) &
```
Comment (gold2718) [inline src/physics/cam/micro_mg_cam.F90]: Isn't this redundant? You only add that pbuf field if `micro_mg_version > 1`. You use this field without the additional check elsewhere in the code so I wonder why put it here?

File: src/physics/cam/micro_mg_cam.F90 [hunk 2]
```diff
@@ -3282,6 +3368,20 @@ subroutine micro_mg_cam_tend_pack(state, ptend, dtime, pbuf, mgncol, mgcols, nle
...
+      qrsedtenout_grid_ptr = qrsedtenout_grid
+   if (qisedten_idx > 0)  qisedtenout_grid_ptr = qisedtenout_grid
+   if (qssedten_idx > 0 .and. micro_mg_version > 1) &
+      qssedtenout_grid_ptr = qssedtenout_grid
+   if (vtrmc_idx > 0)     vtrmcout_grid_ptr    = vtrmcout_grid
+   if (umr_idx > 0 .and. micro_mg_version > 1) &
```
Comment (gold2718) [inline src/physics/cam/micro_mg_cam.F90]: Same redundancy question.

File: src/physics/cam/micro_mg_cam.F90 [hunk 3]
```diff
@@ -3282,6 +3368,20 @@ subroutine micro_mg_cam_tend_pack(state, ptend, dtime, pbuf, mgncol, mgcols, nle
...
+      umrout_grid_ptr      = umrout_grid
+   if (vtrmi_idx > 0)     vtrmiout_grid_ptr    = vtrmiout_grid
+   if (ums_idx > 0 .and. micro_mg_version > 1) &
```
Comment (gold2718) [inline src/physics/cam/micro_mg_cam.F90]: Same redundancy question.

File: src/physics/cam/subcol_SILHS.F90 [hunk 4]
```diff
@@ -1587,12 +1725,2081 @@ function clubb_flip_grid ( profile ) result( profile_flipped )
...
    end function clubb_flip_grid
-
    ! =============================================================================== !
    !                                                                                 !
    ! =============================================================================== !
+#endif
+   !============================================================================
+   subroutine subcol_SILHS_fill_holes_conserv( state, dt, ptend, pbuf )
 
+     ! The William F. Buckley Jr. Conservative Hole Filler.
```
Comment (gold2718) [inline src/physics/cam/subcol_SILHS.F90]: So, radically conservative?

File: src/physics/cam/subcol_SILHS.F90 [hunk 5]
```diff
@@ -1587,12 +1725,2081 @@ function clubb_flip_grid ( profile ) result( profile_flipped )
...
+           enddo ! icol = 1, ncol
+        endif ! any( tot_water_rel_err >= err_thresh )
+
+        ! Print an error message if any total energy relative error is found to
+        ! be greater than the threshold.
+        if ( any( tot_energy_rel_err(:ncol) >= err_thresh ) ) then
+           print *, "Energy conservation error reported in hole filling"
+           do icol = 1, ncol
+              if ( tot_energy_rel_err(icol) >= err_thresh ) then
+                 print *, "Column = ", icol, &
```
Comment (gold2718) [inline src/physics/cam/subcol_SILHS.F90]: This should really be a write to iulog.

File: src/physics/cam/subcol_SILHS.F90 [hunk 6]
```diff
@@ -1587,12 +1725,2081 @@ function clubb_flip_grid ( profile ) result( profile_flipped )
...
+                    - total_energy_column_start(icol) ) ) &
+             / min( total_energy_column_finish(icol), &
+                    total_energy_column_start(icol) )
+
+        enddo ! icol = 1, ncol
+
+        ! Print an error message if any total water relative error is found to
+        ! be greater than the threshold.
+        if ( any( tot_water_rel_err(:ncol) >= err_thresh ) ) then
+           print *, "Water conservation error reported in hole filling"
```
Comment (gold2718) [inline src/physics/cam/subcol_SILHS.F90]: This should really be a write to iulog.

File: src/physics/cam/subcol_SILHS.F90 [hunk 7]
```diff
@@ -1587,12 +1725,2081 @@ function clubb_flip_grid ( profile ) result( profile_flipped )
...
+           do icol = 1, ncol
+              if ( tot_water_rel_err(icol) >= err_thresh ) then
+                 print *, "Column = ", icol, &
```
Comment (gold2718) [inline src/physics/cam/subcol_SILHS.F90]: This should really be a write to iulog.

File: src/physics/cam/subcol_SILHS.F90 [hunk 8]
```diff
@@ -1587,12 +1725,2081 @@ function clubb_flip_grid ( profile ) result( profile_flipped )
...
+                          "Column-integrated grand total water at finish = ", &
+                          grand_total_water_column_finish(icol)
+              endif ! tot_water_rel_err(icol) >= err_thresh
+           enddo ! icol = 1, ncol
+        endif ! any( tot_water_rel_err >= err_thresh )
+
+        ! Print an error message if any total energy relative error is found to
+        ! be greater than the threshold.
+        if ( any( tot_energy_rel_err(:ncol) >= err_thresh ) ) then
+           print *, "Energy conservation error reported in hole filling"
```
Comment (gold2718) [inline src/physics/cam/subcol_SILHS.F90]: This should really be a write to iulog.

File: src/physics/cam/micro_mg_cam.F90 (hunk 1 above)
Comment (brian-eaton) [reply src/physics/cam/micro_mg_cam.F90]: I agree.  The use of the rain and snow pbuf variables is done correctly at other places in the routine.  I'll fix this and the other instances you mention below.

File: src/physics/cam/subcol_SILHS.F90 (hunk 4 above)
Comment (brian-eaton) [reply src/physics/cam/subcol_SILHS.F90]: I had WFB Jr pegged as a traditional conservative:)

Comment (gold2718) [reply src/physics/cam/subcol_SILHS.F90]: I'm going to have to give you this one. In his heyday, he seemed pretty radical to me. Now, he would probably be drummed out of the GOP for being too liberal.
Still my icon for mass conservation is Mr. Rogers. He weighed himself at the same time every day and always came up with the same number.

File: src/physics/cam/subcol_SILHS.F90 (hunk 5 above)
Comment (brian-eaton) [reply src/physics/cam/subcol_SILHS.F90]: Changed all print statements to write to iulog.  Regarding amount of log output, I'd say that this isn't yet production code and I don't know whether triggering the conditionals is rare or common.  We can ask for more control over output if it actually becomes a problem.  I did check the log files from the aux_cam silhs test and didn't find any output from these print statements.

**Comment by brian-eaton** [comment]
The clubb/silhs externals on the svn branch are 
clubb_core/vendor_tags/clubb_ncar_backwards_compat_20181205_c20190528
silhs/vendor_tags/silhs_ncar_backwards_compat_20181205

The current trunk is using 
ESCOMP/CLUBB_CESM tag clubb_ncar_backwards_compat_20181205_c20191001
ESCOMP/SILHS_CESM tag silhs_ncar_backwards_compat_20181205_c20191001

Looks like the cam_development branch and the svn branch are probably the same code version from 20181205.

**Comment by fvitt** [comment]
Don't have time to review all of this.

**Pull request merged**